protocol/tx/entry.go

```diff
@@ -26,10 +26,17 @@ type entry interface {
 var errInvalidValue = errors.New("invalid value")
 
 func entryID(e entry) (hash bc.Hash) {
+	// This nil test only handles the case where e is the zero value of
+	// the entry interface.
 	if e == nil {
 		return hash
 	}
 
+	// This nil test handles the case where e is a nil pointer.
+	if reflect.ValueOf(e).IsNil() {
```

Comment:
Ah I just reread the docs for IsNil; good catch.

I would say in general it's ok (idiomatic and reasonable) to do things that panic as long as the contract is clear to the caller. For example if we have a function that divides, the caller knows they shouldn't provide 0 as the divisor; in that case we don't even need to document that it would panic.

In this case, we would definitely need to document that entryID panics when called with a non-pointer-typed value, or else do the check. And it's not *that* hard to do the check, so why not. We only care about pointer types here, so it would suffice to do:

```go
// Check for both entryID(nil) and entryID(x)
// where x is a concrete nil pointer.
// See https://golang.org/doc/faq#nil_error.
if v := reflect.ValueOf(e); e == nil || v.Kind() == reflect.Ptr && v.IsNil() {
	return bc.Hash{}
}
```